### DISCUSSION TITLE
Consensus on TLSRoute - can we Graduate to Standard?

### DISCUSSION THREAD

**Discussion by candita** [kubernetes-sigs/gateway-api#3410] [General]
The purpose of this discussion is to try to get consensus on 1) keeping TLSRoute and then 2) graduating TLSRoute to standard.

TLSRoute has a well-documented need in Gateway API.  

@hbagdi’s comprehensive document on [[SIG-NETWORK] TLS config in service-apis](https://docs.google.com/document/d/15fkzMrhN_7tA-i2mHKwZpqcjN1o2Pe9Am9Qt828x1lo/edit) provides some background discussion of TLS concerns for service APIs and what it might look like to combine TLS config with HTTPRoute or TCPRoute.   [Service API: TLS related issues](https://docs.google.com/spreadsheets/d/18KE61Y6InCmoQHZcbrYYRZS5Cnt7n33s5dTxUlhHgIA/edit?gid=0#gid=0) lists some Issues that were related back then.  We’ve relied on @youngnick's encyclopedic [Gateway API TLS Use Cases](https://docs.google.com/document/d/17sctu2uMJtHmJTGtBi_awGB0YzoCLodtR6rUNmKMCs8/edit#heading=h.qq2acy7mrk3x) to answer many questions, including the need for TLSRoute when Passthrough routing is required.

The current list of TLS related issues that are waiting to be addressed includes:

- [Support for TLS Certificate Management in Gateway API TLSRoute · Issue #3249](https://github.com/kubernetes-sigs/gateway-api/issues/3249) 
- [TLSRoute promotion to standard #3165](https://github.com/kubernetes-sigs/gateway-api/issues/3165) (includes some of the issues below this)
- [GEP: TLSRoute #2643](https://github.com/kubernetes-sigs/gateway-api/issues/2643)
- [Support TLS Termination mode for TLSRoutes · Issue #2111](https://github.com/kubernetes-sigs/gateway-api/issues/2111)
- [conformance: Coverage for request routing to TLSRoute with no valid BackendRefs · Issue #2153](https://github.com/kubernetes-sigs/gateway-api/issues/2153)
- [Conformance Tests: TLSRoute Passthrough #1579](https://github.com/kubernetes-sigs/gateway-api/issues/1579) 
- [TLS support docs inconsistencies #1474](https://github.com/kubernetes-sigs/gateway-api/issues/1474) 
- [Conflicting SNI's between HTTPRoute & TLSRoute · Issue #623](https://github.com/kubernetes-sigs/gateway-api/issues/623) 

The current list of TLS related discussions expressing interest in TLSRoute includes:

- [TLSRoute match ALPN · Discussion #1683](https://github.com/kubernetes-sigs/gateway-api/discussions/1683)
- [What listeners should be considered when filtering hostnames in routes? · Discussion #1563](https://github.com/kubernetes-sigs/gateway-api/discussions/1563)
- [GatewayTLSConfig: Well known options - Discussion #896](https://github.com/kubernetes-sigs/gateway-api/discussions/896)

Over time [several implementations have added TLSRoute](https://gateway-api.sigs.k8s.io/implementations/v1.1/#tlsroute) based only the experimental API:
Cilium, Contour, Envoy Gateway, Istio, and nginx-gateway-fabric.  I should add that Red Hat needs the TLSRoute (for passthrough) in its Gateway API integrations.

Despite the history, interest expressed via issues and discussions, and several implementations, we find ourselves in a position of questioning the role and necessity of a separate TLSRoute, see [Ideas for TCP, TLS, and UDP Routing · kubernetes-sigs gateway-api · Discussion #2618](https://github.com/kubernetes-sigs/gateway-api/discussions/2618) and [Would Gateway Merging make TLSRoute Unnecessary? · kubernetes-sigs gateway-api · Discussion #3177](https://github.com/kubernetes-sigs/gateway-api/discussions/3177) .  [[SIG-NETWORK] Gateway Hierarchy Brainstorming](https://docs.google.com/document/d/1qj7Xog2t2fWRuzOeTsWkabUaVeOF7_2t_7appe8EXwA/edit#heading=h.w311n4l5qmwk) came about to address a new mechanism for gateway merging and mentions TLSRoute, but without resuming the decision-making on this, it’s not clear that it makes all TLSRoute concerns obsolete.  I would like to request that we try to make a decision on TLSRoute that is not dependent on solving the Gateway Merging issues.  If the community agrees that cannot happen, it would be good to focus energy back on the Gateway Merging solution as soon as possible.

Having said that, here is a summary of reasons to keep TLSRoute:

- How would we otherwise provide that second layer of SNI matching (intersecting with Gateway Listeners when attaching to Gateways).
- How would we otherwise provide a Passthrough solution.
- How would we otherwise provide ALPN routing, mentioned in [GEP: TLS ALPN Routing · Issue #1958](https://github.com/kubernetes-sigs/gateway-api/issues/1958) (unless via GRPCRoute)
- We’d like to keep the persona/roles well-defined. Gateway management should be for cluster admins, but TLSRoute management should be accessible to app developers.

cc @robscott @shaneutt @youngnick @mlavacca

**Comment by shaneutt** [comment]
In addition to the above, I hope we can get representatives from the implementations listed (+Kong, which is not listed but I know to have an implementation) to help us by providing an up-to-date answer to these specific questions:

1. Do you still have need of this API? Do you have active users?
2. Are there any particular difficulties you've found with using (or implementing) this API?
3. Are there things you need from this API that it currently doesn't provide? Are you doing any annotations or custom extensions?

**Comment by mlavacca** [comment]
At Kong, I cannot say `TLSRoute`s a big thing. Usage is pretty limited and we received neither complaints nor requests for improvements so far.  Also, I think we need to address the gateway merging problem, as that's something that affects many implementations in one way or the other and hinders multiple proposals from being moved on.  

However, after having discussed Gateway merging and thinking about it a lot recently, my feeling is that `TLSRoute`s development should not be blocked by it. As @candita said, there are multiple reasons to move forward `TLSRoute`s, and I agree with them. I personally like the approach of Gateway merging described in https://github.com/kubernetes-sigs/gateway-api/pull/3213, but I do not think this approach should aim at replacing L4 routes, mostly because this would break our persona model - app developers should be exposed to another layer of the API, which is something I think we should avoid.

**Comment by youngnick** [comment]
Cilium definitely has usage of TLSRoute, and we receive questions about when it will move forward fairly regularly.

**Comment by rikatz** [comment]
Given TLSRoute is graduated, I am closing this discussion. Thanks all!
